internal/controller/servermaintenanceset_controller.go

```diff
@@ -0,0 +1,247 @@
+	"github.com/ironcore-dev/controller-utils/metautils"
+	metalv1alpha1 "github.com/ironcore-dev/metal-operator/api/v1alpha1"
+)
+
+// ServerMaintenanceSetFinalizer is the finalizer for ServerMaintenanceSet
+const ServerMaintenanceSetFinalizer = "metal.ironcore.dev/servermaintenanceset"
+
+// ServerMaintenanceSetReconciler reconciles a ServerMaintenanceSet object
+type ServerMaintenanceSetReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=metal.ironcore.dev,resources=servermaintenancesets,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=metal.ironcore.dev,resources=servermaintenancesets/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=metal.ironcore.dev,resources=servermaintenancesets/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the ServerMaintenanceSet object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.20.2/pkg/reconcile
+func (r *ServerMaintenanceSetReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	replica := &metalv1alpha1.ServerMaintenanceSet{}
+
+	if err := r.Get(ctx, req.NamespacedName, replica); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	return r.reconcileExists(ctx, replica)
+}
+
+func (r *ServerMaintenanceSetReconciler) reconcileExists(ctx context.Context, replica *metalv1alpha1.ServerMaintenanceSet) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+	if !replica.DeletionTimestamp.IsZero() {
+		return r.delete(ctx, log, replica)
+	}
+
+	if modified, err := clientutils.PatchEnsureFinalizer(ctx, r.Client, replica, ServerMaintenanceSetFinalizer); err != nil || modified {
+		return ctrl.Result{}, err
+	}
+	log.V(1).Info("Ensured finalizer has been added")
+
+	return r.reconcile(ctx, log, replica)
+}
+
+func (r *ServerMaintenanceSetReconciler) delete(ctx context.Context, log logr.Logger, replica *metalv1alpha1.ServerMaintenanceSet) (ctrl.Result, error) {
+	log.V(1).Info("Deleting ServerMaintenanceSet")
+
+	maintenancelist, err := r.getOwnedMaintenances(ctx, replica)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	for _, maintenance := range maintenancelist.Items {
+		if maintenance.Status.State == metalv1alpha1.ServerMaintenanceStateInMaintenance ||
+			maintenance.Status.State == metalv1alpha1.ServerMaintenanceStateCompleted {
+			log.V(1).Info("Owned Maintenance is still inMaintenance or completed", "maintenance", maintenance.Name)
+			continue
+		}
+		if err := r.Delete(ctx, &maintenance); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
```

Comment:
So we are setting the ownerref on the `ServerMaintenance` objects when we create them. Isn't that contradicting with what the GC of kube will do - setting the `DeletionTimeStap` on those mgmt resources?

Reply:
Do you mean this controller should not take care of the deletion, just let the GC do it? 
I think we discussed this with Eric and I was told to call Delete on those maintenance CRDs instead of relying on the GC.
But I am ok with just letting GC handling it

Reply:
@afritzler fyi